sailfishusermanager.cpp

```diff
@@ -0,0 +1,429 @@
+/*
+ * Copyright (c) 2020 Jolla Ltd.
+ * Copyright (c) 2020 Open Mobile Platform LLC.
+ *
+ * All rights reserverd.
+ *
+ * BSD 3-Clause License, see LICENSE.
+ */
+
+#include "sailfishusermanager.h"
+#include "usermanager_adaptor.h"
+
+#include <QDBusConnection>
+#include <QTimer>
+#include <QFile>
+#include <QDir>
+
+#include <pwd.h>
+#include <grp.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <libuser/user.h>
+
+const char* DBUS_INTERFACE = "org.sailfishos.usermanager";
+const char* DBUS_OBJECT_PATH = "/";
+const char* USER_GROUP = "users";
+const char* GROUPS_IDS_FILE = "/usr/lib/setup/group.ids";
+const char* GROUPS_SYSTEM = "SYSTEM_GROUPS";
+const char* GROUPS_USER = "USER_GROUPS";
+const char* SKEL_DIR = "/etc/skel";
+const char* USER_HOME = "/home/%1";
+const int HOME_MODE = 0700;
+const int QUIT_TIMEOUT = 60 * 1000; // One minute quit timeout
+
+SailfishUserManager::SailfishUserManager(QObject *parent) : QObject(parent)
+{
+    qDBusRegisterMetaType<SailfishUser>();
+    qDBusRegisterMetaType<QList<SailfishUser>>();
+
+    QDBusConnection connection = QDBusConnection::systemBus();
+    if (!connection.registerObject(DBUS_OBJECT_PATH, this)) {
+        qFatal("Cannot register D-Bus object at %s", DBUS_OBJECT_PATH);
+    }
+
+    if (!connection.registerService(DBUS_INTERFACE)) {
+        qFatal("Cannot register D-Bus service at %s", DBUS_INTERFACE);
+    }
+
+    new UsermanagerAdaptor(this);
+
+    m_timer = new QTimer(this);
+    connect(m_timer, SIGNAL(timeout()), qApp, SLOT(quit()));
+    m_timer->start(QUIT_TIMEOUT);
+}
+
+QList<SailfishUser> SailfishUserManager::users(const QDBusMessage& message)
+{
+    m_timer->start();
+    QList<SailfishUser> rv;
+
+    struct group *grent = getgrnam(USER_GROUP);
+    if (grent) {
+        for (int i = 0; grent->gr_mem[i]; i++) {
+            SailfishUser user;
+            user.user = grent->gr_mem[i];
+
+            struct passwd *pw = getpwnam(grent->gr_mem[i]);
+            if (pw && pw->pw_gecos)
+                user.name = pw->pw_gecos;
+
+            rv.append(user);
+        }
+    } else {
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Getting user group failed"));
+    }
+
+    return rv;
+}
+
+bool SailfishUserManager::addGroup(const QString &group)
+{
+    struct lu_error *error = nullptr;
+    struct lu_context *context = lu_start(NULL, lu_user, NULL, NULL, lu_prompt_console, NULL, &error);
+    bool rv = true;
+
+    if (!context) {
+        qWarning() << "Error creating context:" << lu_strerror(error);
+        return false;
+    }
+
+    struct lu_ent *ent_group = lu_ent_new();
+    if (lu_group_default(context, group.toLocal8Bit(), false, ent_group)) {
+        if (!lu_group_add(context, ent_group, &error)) {
+            qWarning() << "Adding group failed:" << lu_strerror(error);
```

Comment:
There is lu_error_free() function, but no calls to it. Am I missing something or are these memory leaks?